src/openms/source/CHEMISTRY/Tagger.cpp

```diff
@@ -0,0 +1,201 @@
+      tag.push_back(aa);
+      getTag_(tag, mzs, j, tags);
+      // each tag is build as a StringList, so that push_back(), pop_back() and size() work as intended even with modified residues, e.g. "M(Oxidation)"
+      // and then the tag is reported as a string by collapsing it here using concatenate()
+      if (tag.size() >= min_tag_length_) tags.insert(ListUtils::concatenate(tag));
+
+      // if aa is "L", then also add "I" as an alternative residue and extend the tag again
+      // this will add redundancy, (and redundant runtime) but we avoid dealing with J and ambigous matching to I and L later on
+      // TODO would it be more efficient to only generate tags with L and after they are finished make copies with L exchanged for I?
+      // TODO Then we would need to generate all possible combinations, if multiple L are in the tag
+      if (aa == "L")
+      {
+        tag.pop_back();
+        tag.push_back("I");
+        getTag_(tag, mzs, j, tags);
+      }
+      if (tag.size() >= min_tag_length_) tags.insert(ListUtils::concatenate(tag));
+      tag.pop_back();  // remove last string
+      ++j;
+    }
+  }
+
+  Tagger::Tagger(size_t min_tag_length, float ppm, size_t max_tag_length, size_t min_charge, size_t max_charge, StringList fixedModNames, StringList varModNames)
+  {
+    ppm_ = ppm;
+    min_tag_length_ = min_tag_length;
+    max_tag_length_ = max_tag_length;
+    min_charge_ = min_charge;
+    max_charge_ = max_charge;
+
+    const std::set<const Residue*> aas = ResidueDB::getInstance()->getResidues();
+
+    for (const auto& r : aas)
+    {
+      std::string letter;
+      letter += r->getOneLetterCode()[0];
+
+      // skip non-standard residues, could be made into a parameter (TODO?)
+      // "I" is skipped here and handled explicitly as an alternative to "L" in getTag_
+      if (letter == "B" || letter == "X" || letter == "Z" || letter == "O" || letter == "J" || letter == "I")
+      {
+        continue;
+      }
+      const float mass = r->getMonoWeight(Residue::Internal);
+      mass2aa[mass] = letter;
+      std::cout << "TEST TAGGER added residue: " << letter << " | mass: " << mass << std::endl;
+    }
+
+    // for fixed modifications, replace the unmodified residue with the modified one
+    for (String mod : fixedModNames)
+    {
+      const ResidueModification* rm = ModificationsDB::getInstance()->getModification(mod);
+      std::cout << "TEST TAGGER rm.getOrigin(): " << rm->getOrigin() << " | rm.getName(): " << rm->getName() << " | rm.getId(): " << rm->getId() << std::endl;
+      Residue r = *(ResidueDB::getInstance()->getResidue(rm->getOrigin()));
+      std::cout << "TEST TAGGER unmodified residue mass: " << r.getMonoWeight(Residue::Internal) << std::endl;
+      r.setModification(rm->getId());
+      std::cout << "TEST TAGGER modified residue mass: " << r.getMonoWeight(Residue::Internal) << std::endl;
+
+      for (std::map<float, std::string>::iterator it = mass2aa.begin(); it != mass2aa.end(); ++it)
```

Comment:
this is an inverse map.. and ineffective. But does not hurt I guess, because the number of items is small?